### ISSUE TITLE
Kernel: First killed process is never removed from all processes list

### ISSUE THREAD

**Issue by ADKaster** [SerenityOS/serenity#4337]
This one's pretty weird.

Steps:
0) Apply this patch for log spam (from resource graph, which polls /proc/all)
```
diff --git a/Kernel/FileSystem/ProcFS.cpp b/Kernel/FileSystem/ProcFS.cpp
index c81b41005..c98266446 100644
--- a/Kernel/FileSystem/ProcFS.cpp
+++ b/Kernel/FileSystem/ProcFS.cpp
@@ -886,6 +886,9 @@ static Optional<KBuffer> procfs$all(InodeIdentifier)
             thread_object.add("times_scheduled", thread.times_scheduled());
             thread_object.add("ticks", thread.ticks());
             thread_object.add("state", thread.state_string());
+            if (thread.state() == Thread::State::Dead) {
+                dbgln("procfs$all: Why is dead tid {} from {}({}) still here?", thread.tid().value(), process.name(), process.pid().value());
+            }
             thread_object.add("cpu", thread.cpu());
             thread_object.add("priority", thread.priority());
             thread_object.add("effective_priority", thread.effective_priority());
```

1) Start terminal
2) Start any other application (e.g. FileManager)
3) `ps -e` # find pid of app you started
4) `kill -9 <pid>`
5) What?
```
[FileManager(24:24)]: Terminating FileManager(24) due to signal 9
[Shell(22:22)]: select was interrupted
Shell(22): Job entry "kill -9 24" deleted in 40 ms
[ResourceGraph.MenuApplet(9:9)]: procfs$all: Why is dead tid 24 from FileManager(24) still here?
[ResourceGraph.MenuApplet(9:9)]: procfs$all: Why is dead tid 24 from FileManager(24) still here?
[ResourceGraph.MenuApplet(9:9)]: procfs$all: Why is dead tid 24 from FileManager(24) still here?
[ResourceGraph.MenuApplet(9:9)]: procfs$all: Why is dead tid 24 from FileManager(24) still here?
[ResourceGraph.MenuApplet(9:9)]: procfs$all: Why is dead tid 24 from FileManager(24) still here?
```

More observations:
- The GUI for the killed app will still be present, but 100% unresponsive. Which makes sense, it's a process with a dead shell of an app in it. No active threads.
- Killing more processes after the first one behave as expected. They are removed from the list of all processes and do not show up in /proc/all.

**Comment by ADKaster** [comment]
Ok, interesting fact. If I add these methods to Process.h
```c++
    void will_be_destroyed() const
    {
        dbgln("{}({}) will be destroyed now!", m_name, m_pid.value());
    }
    void one_ref_left() const
    {
        dbgln("{}({}) has one reference left now!", m_name, m_pid.value());
    }
```


Then the results are a bit more enlightening.

```
[FileManager(25:25)]: Terminating FileManager(25) due to signal 15
[FinalizerTask(3:3)]: FileManager(25) has one reference left now!
[FinalizerTask(3:3)]: kill(26) has one reference left now!
[Shell(22:22)]: select was interrupted
[Shell(22:22)]: kill(26) has one reference left now!
[Shell(22:22)]: kill(26) will be destroyed now!
Shell(22): Job entry "kill 25" deleted in 43 ms
[ResourceGraph.MenuApplet(9:9)]: FileManager(25) has one reference left now!
```
In the finalizer, there is a reference count remaining for the process. Who is holding onto it? The world may never know :)
Also surprising is that procfs$all takes a strong reference to every single process. Would it be better for it to take a vector of WeakPtr instead?

**Comment by ADKaster** [comment]
Oh, the thread is holding onto it. So the real question is, who is keeping the thread alive in this case?

**Comment by tomuta** [comment]
Probably related to the changes I made in how blocking works which is drastically different. I'll look into it hopefully this weekend.

**Comment by tomuta** [comment]
If you open SystemMonitor it becomes slightly more obvious. Basically, FileManager has two threads, one of which got killed. But it looks like the other thread is deadlocked trying to re-acquire the process lock (which the dead thread probably didn't release I'm guessing), and my guess is that if it weren't for that it would continue on and kill itself shortly after acquiring the lock.

**Comment by tomuta** [comment]
I think this one will also be resolved by #4347

**Comment by ADKaster** [comment]
This was resolved a while back but never closed.
